cmd/kubeops/internal/handler/dashboardcompat_test.go

```diff
@@ -0,0 +1,200 @@
+						"password": "123456",
+						"name":     "user1",
+					},
+					"port": 3000,
+				},
+			},
+			Helm2Release: h2.Release{
+				Name:      "Foo",
+				Namespace: "default",
+				Info: &h2.Info{
+					Status: &h2.Status{
+						Code: h2.Status_DEPLOYED,
+					},
+				},
+				Chart: &h2chart.Chart{
+					Metadata: &h2chart.Metadata{},
+					Values: &h2chart.Config{
+						Raw: "{}\n",
+					},
+				},
+				Version: 1,
+				Config: &h2chart.Config{
+					Raw: "port: 3000\nuser:\n  name: user1\n  password: \"123456\"\n",
+				},
+			},
+		},
+		{
+			Description:         "Two equivalent releases with different versions",
+			ShouldFail:          true,
+			MarshallingFunction: asResponse,
+			Helm3Release: h3.Release{
+				Name:      "Foo",
+				Namespace: "default",
+				Chart: &h3chart.Chart{
+					Metadata: &h3chart.Metadata{},
+					Values: map[string]interface{}{
+						"port": 8080,
+					},
+				},
+				Info: &h3.Info{
+					Status: h3.StatusDeployed,
+				},
+				Version: 1,
+				Config: map[string]interface{}{
+					"port": 3000,
+					"user": map[string]interface{}{
+						"name":     "user1",
+						"password": "123456",
+					},
+				},
+			},
+			Helm2Release: h2.Release{
+				Name:      "Foo",
+				Namespace: "default",
+				Info: &h2.Info{
+					Status: &h2.Status{
+						Code: h2.Status_DEPLOYED,
+					},
+				},
+				Chart: &h2chart.Chart{
+					Metadata: &h2chart.Metadata{},
+					Values: &h2chart.Config{
+						Raw: "port: 8080\n",
+					},
+				},
+				Version: 5,
+				Config: &h2chart.Config{
+					Raw: "port: 3000\nuser:\n  name: user1\n  password: \"123456\"\n",
+				},
+			},
+		},
+		{
+			Description:         "Incomplete Helm 3 Release",
+			ShouldFail:          true,
+			MarshallingFunction: asResponse,
+			Helm3Release:        h3.Release{Name: "Incomplete"},
+			Helm2Release:        h2.Release{Name: "Incomplete"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.Description, func(t *testing.T) {
+			// A panic is a failure for the test
+			defer func() {
+				if r := recover(); r != nil {
+					if !test.ShouldFail {
+						t.Errorf("Not expected to fail, yet got a panic: %v. \nStacktrace: \n%s", r, string(debug.Stack()))
+					}
+				}
+			}()
+			// Perform conversion
+			compatibleH3rls := newDashboardCompatibleRelease(test.Helm3Release)
+			// Marshall both: Compatible H3Release and H2Release
+			h3Marshalled := test.MarshallingFunction(compatibleH3rls)
+			t.Logf("Marshalled Helm 3 Release %s", h3Marshalled)
+			h2Marshalled := test.MarshallingFunction(test.Helm2Release)
+			t.Logf("Marshalled Helm 2 Release %s", h2Marshalled)
+			// Check result
+			areEqual := h2Marshalled == h3Marshalled
+			if test.ShouldFail && areEqual {
+				t.Errorf("Expected to fail, but are equal %v", cmp.Diff(h3Marshalled, h2Marshalled))
+			}
+			if !test.ShouldFail && !areEqual {
+				t.Errorf("Not equal: %s", cmp.Diff(h3Marshalled, h2Marshalled))
+			}
+		})
+	}
+}
+
+func asResponse(data h2.Release) string {
```

Comment:
can you explain why you are using `asResponse`? Why not using directly a marshall function?

Reply:
still not sure about this ^

Reply:
In the current setting, the result from the function `newDashboardCompatibleRelease` is used in `kubeops/internal/handler` [here](https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeops/internal/handler/handler.go#L170) & [here](https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeops/internal/handler/handler.go#L183), which both marshal the data using `github.com/unrolled/render`. To avoid digging in the aforementioned library _and_ to enable further tests where `newDashboardCompatibleRelease` are marshalled/viewed in different ways, I created `asResponse` as a helper function to mimic the behaviour in `/cmd/kubeops/internal/handler/handler.go`.

If my reasoning seems too stretched I can just bake it into the test.

Reply:
That's fine, a comment explaining that would be helpful for the future